### PR TITLE
Simplify bringup

### PR DESCRIPTION
**Pull request by SteveMacenski**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | https://github.com/ros-planning/navigation2/issues/1216 |
| Primary OS tested on | (Ubuntu, MacOS, Windows) |
| Robotic platform tested on | (Steve's Robot, gazebo simulation of Tally, hardware turtlebot) |

---

## Description of contribution in a few bullet points

* Simplifying bringup. No need to declare a bunch of parameters the later launch files are going to declare and use
* Give the launch files control over their own node management. There's not really much reason to have 1 lifecycle manager globally when there's 2 main launch files that can maintain themselves (and now 1 list of each node)

Something I noticed though, and I see this on `master` so not related to this PR, is that I'm having problems reading in the parameters file right now. I'd like it if someone else were to run off master & tell me if they're having issues. If I just run out of the box, I get a load error for `declare_parameter` in `map_server` for `yaml_filename` because its not taking the override. Same thing for the bt navigator `plugin_lib_names` libraries, its taking the default `nav2_behavior_tree_nodes` which no longer exists (and that default was removed)

This has been happening to me on and off since Crystal and I've yet to be able to figure out why that is, but its worse right now than normal so I'm actually wondering if master is broken.

## Future work

* the nav2 simulation launch files can have their excessive declarations removed as well

### CODE REVIEW

**Comment by crdelsey** [comment]
I'm seeing a different error when running master, but possibly related:
```
[planner_server-8] terminate called after throwing an instance of 'rclcpp::exceptions::UnknownROSArgsError'
[planner_server-8]   what():  found unknown ROS arguments: '', '', '', '', '', '', '', ''
```

I get something similar for pretty much every node.

**Comment by SteveMacenski** [comment]
What's your version of ROS master? I haven't updated since Foxy. So that is probably unrelated

**Comment by SteveMacenski** [comment]
OK I still see this on master on a clean eloquent build. I think there's something wrong with master but not related to this PR

**Comment by SteveMacenski** [comment]
Resolved, it was my bad mixed in with a dirty build environment.

**Comment by mkhansen-intel** [comment]
Steve if you are seeing issues unrelated to this PR on master, can you file an issue? I'm a little confused from your monologue above.

File: nav2_bringup/bringup/launch/nav2_bringup_launch.py [hunk 1]
```diff
@@ -111,21 +109,6 @@ def generate_launch_description():
                               'use_lifecycle_mgr': 'false',
                               'use_remappings': use_remappings,
                               'map_subscribe_transient_local': 'true'}.items()),
-
```
Comment (crdelsey) [inline nav2_bringup/bringup/launch/nav2_bringup_launch.py]: This change will result in two instances of lifecycle_manager getting launched, one managing amcl and map server and the other managing the rest of the nodes.

Both lifecycle_managers will have the same node name and be in the same namespace, so they will not be individually addressable. In the case of `autostart=False`, there is no way to reliably address the specific lifecycle managers to cause them to start lifecycle transitions.

The `nav2_tb3_simulation_launch.py` starts the system with `autostart=False` and relies on the RViz plugin to perform the lifecycle transitions. Also the RViz plugin is unaware that it needs to talk to two lifecycle managers now.

There are 3 approaches I can see to address this problem:
1. Leave it as it was
1. Put the two lifecycle managers in separate namespaces, always autostart, and remove lifecycle_manager interaction from the Rviz plugin.
3. Put the two lifecycle managers in separate namespaces, and add code to find all lifecycle_manager services in the system, and start them all when the startup button is pressed.
4. Change the way lifecycle manager works so each launch file can add it's own nodes to the startup list.

Comment (SteveMacenski) [reply nav2_bringup/bringup/launch/nav2_bringup_launch.py]: Just going inline:

> This change will result in two instances of lifecycle_manager getting launched, one managing amcl and map server and the other managing the rest of the nodes.

Agreed.

> Both lifecycle_managers will have the same node name and be in the same namespace, so they will not be individually addressable. 

That is not true, they have different names, `lifecycle_manager_navigation` and `lifecycle_manager_localization`.

> In the case of autostart=False, there is no way to reliably address the specific lifecycle managers to cause them to start lifecycle transitions.

That would be true, I hadn't thought about that. 

If these are truly intended to be independent launch files, I think its totally reasonable that they each have their own lifecycle manager, or we need a single global-scope one that is given all the nodes to launch. The benefit of this is that a single launch file contains everything it needs to be successful. If you want to run your navigation with mapping, kill mapping, and move to localization, you can do that without effecting the navigation's lifecycle. That's very beneficial to me. I wouldn't like the idea of a navigation-base launch file without being able to transition fully into active. I think you could optionally dump the AMCL/map server stuff into that same lifecycle manager, but its not clear to me how to do that in the current launch system.

If we chose to leave it as it was, then I think the ticket should be closed, because that is the scope of the ticket, to remove redundancy in the lifecycle node declarations. I'd prefer to default everything to auto-bringup or add it to the rviz plugin, but that doesn't seem like a good move to have multiple, possibly there, possibly not lifecycle manager clients depending on if you're in slam or localization mode. 

Thoughts?

Comment (SteveMacenski) [reply nav2_bringup/bringup/launch/nav2_bringup_launch.py]: Another point to mention, those other 2 lifecycle managers were inside the launch files, I just made them default and removed the single one. If the rviz plugin or multiple managers is an issue, those then need to just be flat out removed.

Comment (crdelsey) [reply nav2_bringup/bringup/launch/nav2_bringup_launch.py]: > I think its totally reasonable that they each have their own lifecycle manager

I'm ok with that, but there was some resistance to having two separate lifecycle managers running, back when these launch files were rewritten.  @mkhansen-intel Were you the one who preferred have one lifecycle manager running instead of two?

If we go the route of having two lifecycle managers, I'm ok with making autostart=True the default in all our launch files. For the Rviz plugin, I'm inclined to pull the lifecycle stuff out of it and making the lifecycle management a separate tool

Comment (SteveMacenski) [reply nav2_bringup/bringup/launch/nav2_bringup_launch.py]: Or, we just hardcode in lifecycle managers startup and just entirely abstract out the concept to users. 

I don't mind the button, but it does make it so that any modification to the example launch files basically makes the rviz plugin stop functioning, which isn't very bulletproof.

More thoughts would be good on this before going one way or another. Maybe a discussion point for tomorrow.

Comment (crdelsey) [reply nav2_bringup/bringup/launch/nav2_bringup_launch.py]: I think we're on the same page. I was thinking the separate lifecycle management tool would be outside of RViz and completely optional if the user wants to use our system lifecycle transitions.

I think the long-term goal is to migrate lifecycle manager into `launch`. This seems like a step in that direction.

Comment (SteveMacenski) [reply nav2_bringup/bringup/launch/nav2_bringup_launch.py]: Well Launch has a way to transition into lifecycle active states. Its a mess, but it exists. Though I haven't figured out how to transition to shutdown on control-C, but I'm not convinced its _not_ possible, but challenging. Long-term reimplementing the lifecycle manager as a python launch system item sounds interesting. Maybe the `LifecycleNode()` has another flag for `add_to_manager` and in the background it aggregates the _ordered_ lifecycle nodes and brings them all up in that fashion. Still need to figure out how to handle the `SIG`s to transition to shutdown before exiting. This sounds like a really good future work item. I'll file a ticket as a reminder. Maybe I'll have an intern eventually who can do that.

The rviz plugin panel really only deals with the lifecycle states and pausing navigation. I think that's an appropriate location for it, though I'd recommend automatically starting up by default in all launch files. The less friction for users to get this going the better. Plus if they dont want to use rviz, that will make it very hard for them to find out about the activation stuff (well, I guess other than the documentation ;-) ).

**Comment by SteveMacenski** [comment]
@crdelsey I changed the default to bringup on startup, and changed the rviz plugin to operate with the localization launch file as well.

I'm having now 2 problems: 
1. I'm seeing this in master too: I can't seem to use the set pose tool in rviz to create the map->odom transform. I see this problem here too. I'm investigating.
2. If in SLAM mode, the rviz plugin will fail to work. I couldn't add a second button because of the state machine design would make it incomprehensibly hard to follow. My solution to this is to do one of the following:
  - As you suggested in an earlier post, remove the lifecycle startup button all together if we have it auto start
  - Make the localization bringup best effort, such that it will try to bring up, but if the services associated with it `/is_active` and `/manage_nodes` don't exist, we'll just skip it because its not around

Thoughts?

**Comment by SteveMacenski** [comment]
My last commit, 35922c6cbb70a3b359ef90cdc733bc84c3b19f4f, accomplishes my preferred action, but its isolated to that commit so it can be easily rolled back if you like another route.

I did a clean uninstall and reinstall of ROS and I'm getting things to work again with the map->odom transform. That was weird.

Ready for review again.

File: nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp [hunk 2]
```diff
@@ -146,17 +147,34 @@ class InitialThread : public QThread
...
+    bool tried_loc_bringup_once = false;
+    while (status_loc == SystemStatus::TIMEOUT) {
+      status_loc = client_loc_.is_active(std::chrono::seconds(1));
+      if (tried_loc_bringup_once) {
+        break;
+      }
+      tried_loc_bringup_once = true;
     }
-    if (status == SystemStatus::ACTIVE) {
+
```
Comment (mkhansen-intel) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp]: So the user has no way of knowing that localization hasn't been started up, only if navigation startup succeeded, right?

Seems like an enhancement is needed to the panel to show status for both

Comment (SteveMacenski) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp]: There are logging statements, so there should be some knowledge.

Comment (mkhansen-intel) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp]: If the user is in the RVIZ GUI on a remote machine, they won't see a message in the terminal, right?

I won't block this PR based on this but could we get an enhancement to the nav2 Rviz panel to show status for both in a follow on PR?

Comment (SteveMacenski) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp]: The error is appearing in the RVIZ terminal.

If your argument is that we can't expect someone to see terminal output, then we shouldn't include any at all anywhere. This isn't unique to this situation.

I think that's a reasonable request -- I'll file a ticket.

Comment (mkhansen-intel) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp]: You're right, I don't know why I was thinking it would appear in the server's terminal. I think we're in agreement now.

**Comment by mkhansen-intel** [comment]
I approve

**Comment by SteveMacenski** [comment]
Letting CI do its thing then merging with Matt's approval
